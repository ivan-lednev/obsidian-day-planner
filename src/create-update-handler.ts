import type { App } from "obsidian";
import { isNotVoid } from "typed-assert";

import { sortListsRecursivelyInMarkdown } from "./mdast/mdast";
import {
  createTransaction,
  getTaskDiffFromEditState,
  mapTaskDiffToUpdates,
  TransactionWriter,
} from "./service/diff-writer";
import type { PeriodicNotes } from "./service/periodic-notes";
import type { VaultFacade } from "./service/vault-facade";
import type { DayPlannerSettings } from "./settings";
import type { OnUpdateFn } from "./types";
import { type ConfirmationModalProps } from "./ui/confirmation-modal";
import { EditMode } from "./ui/hooks/use-edit/types";
import { SingleSuggestModal } from "./ui/SingleSuggestModal";
import { applyScopedUpdates } from "./util/markdown";

export async function getTextFromUser(app: App): Promise<string | undefined> {
  return new Promise((resolve) => {
    new SingleSuggestModal({
      app,
      getDescriptionText: (value) =>
        value.trim().length === 0
          ? "Start typing to create a task"
          : `Create item "${value}"`,
      onChooseSuggestion: async ({ text }) => {
        resolve(text);
      },
      onClose: () => {
        resolve(undefined);
      },
    }).open();
  });
}

export const createUpdateHandler = (props: {
  settings: () => DayPlannerSettings;
  transactionWriter: TransactionWriter;
  vaultFacade: VaultFacade;
  periodicNotes: PeriodicNotes;
  onEditCanceled: () => void;
  onEditConfirmed: () => void;
  getTextInput: () => Promise<string | undefined>;
  getConfirmationInput: (input: ConfirmationModalProps) => Promise<boolean>;
}): OnUpdateFn => {
  const {
    settings,
    transactionWriter,
    vaultFacade,
    onEditCanceled,
    onEditConfirmed,
    periodicNotes,
    getTextInput,
    getConfirmationInput,
  } = props;

  function getPathsToCreate(paths: string[]) {
    return paths.reduce<string[]>(
      (result, path) =>
        vaultFacade.checkFileExists(path) ? result : result.concat(path),
      [],
    );
  }

  return async (base, next, mode) => {
    const diff = getTaskDiffFromEditState(base, next);

    if (mode === EditMode.CREATE) {
      const created = diff.added[0];

      isNotVoid(created);

      const modalOutput = await getTextInput();

      if (!modalOutput) {
        return onEditCanceled();
      }

      diff.added[0] = { ...created, text: modalOutput };
    }

    const updates = mapTaskDiffToUpdates(diff, mode, settings(), periodicNotes);

    const afterEach = settings().sortTasksInPlanAfterEdit
      ? (contents: string) =>
          applyScopedUpdates(
            contents,
            settings().plannerHeading,
            sortListsRecursivelyInMarkdown,
          )
      : undefined;

    const transaction = createTransaction({
      updates,
      afterEach,
      settings: settings(),
    });

    const updatePaths = [...new Set([...transaction.map(({ path }) => path)])];

    const needToCreate = getPathsToCreate(updatePaths);

    if (needToCreate.length > 0) {
      const confirmed = await getConfirmationInput({
        title: "Need to create files",
        text: `The following files need to be created: ${needToCreate.join("; ")}`,
        cta: "Create",
      });

      if (!confirmed) {
        return onEditCanceled();
      }

      await Promise.all(
        needToCreate.map(async (path) => {
          const date = periodicNotes.getDateFromPath(path, "day");

          isNotVoid(date);

          await periodicNotes.createDailyNote(date);
        }),
      );
    }

    await transactionWriter.writeTransaction(transaction);

    return onEditConfirmed();
  };
};
