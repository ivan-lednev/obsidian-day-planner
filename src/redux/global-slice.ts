import { createSelector, type PayloadAction } from "@reduxjs/toolkit";

import { defaultDayFormat } from "../constants";

import { createAppSlice } from "./create-app-slice";

interface ObsidianSliceState {
  visibleDays: string[];
}

export const initialState: ObsidianSliceState = {
  visibleDays: [],
};

export const globalSlice = createAppSlice({
  name: "obsidian",
  initialState,
  reducers: (create) => ({
    visibleDaysUpdated: create.reducer(
      (state, action: PayloadAction<string[]>) => {
        state.visibleDays = action.payload;
      },
    ),
    editCanceled: () => {},
  }),
  selectors: {
    selectVisibleDays: (state) => state.visibleDays,
  },
});

export const { visibleDaysUpdated, editCanceled } = globalSlice.actions;

export const { selectVisibleDays } = globalSlice.selectors;

export const selectSortedDedupedVisibleDays = createSelector(
  selectVisibleDays,
  (days) => {
    return [...new Set(days)]
      .sort()
      .map((it) => window.moment(it, defaultDayFormat));
  },
);
