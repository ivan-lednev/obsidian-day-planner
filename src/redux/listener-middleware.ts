import { createListenerMiddleware } from "@reduxjs/toolkit";

import { icalParseLowerLimit } from "../constants";
import type { ReduxExtraArgument } from "../types";
import { createBackgroundBatchScheduler } from "../util/scheduler";

import { dataviewChange } from "./dataview/dataview-slice";
import { createListPropsParseListener } from "./dataview/init-dataview-listeners";
import { icalRefreshRequested, icalsFetched } from "./ical/ical-slice";
import {
  checkIcalEventsChanged,
  checkVisibleDaysChanged,
  createCachingFetcher,
  createIcalFetchListener,
  createIcalParseListener,
  type IcalParseTaskResult,
} from "./ical/init-ical-listeners";
import type { AppDispatch, RootState } from "./store";

export function initListenerMiddleware(props: { extra: ReduxExtraArgument }) {
  const {
    extra,
    extra: { listPropsParser },
  } = props;

  const icalParseScheduler =
    createBackgroundBatchScheduler<IcalParseTaskResult>({
      timeRemainingLowerLimit: icalParseLowerLimit,
    });

  const listenerMiddleware = createListenerMiddleware<
    RootState,
    AppDispatch,
    ReduxExtraArgument
  >({
    extra,
  });

  listenerMiddleware.startListening({
    actionCreator: icalRefreshRequested,
    effect: createIcalFetchListener({ fetcher: createCachingFetcher() }),
  });

  listenerMiddleware.startListening({
    predicate: (action, currentState) =>
      checkIcalEventsChanged(action, currentState) ||
      checkVisibleDaysChanged(action, currentState),
    effect: createIcalParseListener({
      scheduler: icalParseScheduler,
    }),
  });

  listenerMiddleware.startListening({
    actionCreator: icalsFetched,
    effect: async (action, listenerApi) => {
      await listenerApi.extra.onIcalsFetched(action.payload);
    },
  });

  listenerMiddleware.startListening({
    actionCreator: dataviewChange,
    effect: createListPropsParseListener({ listPropsParser }),
  });

  return listenerMiddleware;
}
