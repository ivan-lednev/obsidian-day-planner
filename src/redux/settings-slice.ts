import { type PayloadAction } from "@reduxjs/toolkit";

import { type DayPlannerSettings, defaultSettings } from "../settings";

import { createAppSlice } from "./create-app-slice";

interface SettingsSliceState {
  settings: DayPlannerSettings;
}

export const initialState: SettingsSliceState = {
  settings: defaultSettings,
};

export const settingsSlice = createAppSlice({
  name: "settings",
  initialState,
  reducers: (create) => ({
    settingsUpdated: create.reducer(
      (state, action: PayloadAction<Partial<DayPlannerSettings>>) => {
        state.settings = { ...state.settings, ...action.payload };
      },
    ),
    settingsLoaded: create.reducer(
      (state, action: PayloadAction<DayPlannerSettings>) => {
        state.settings = action.payload;
      },
    ),
  }),
  selectors: {
    selectDataviewSource: (state) => state.settings.dataviewSource,
    selectIcals: (state) => state.settings.icals,
    selectSettings: (state) => state.settings,
  },
});

export const { settingsUpdated, settingsLoaded } = settingsSlice.actions;

export const { selectDataviewSource, selectSettings, selectIcals } =
  settingsSlice.selectors;
