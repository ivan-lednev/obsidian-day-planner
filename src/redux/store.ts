import type {
  Action,
  ConfigureStoreOptions,
  ListenerEffect,
  ListenerMiddlewareInstance,
  ThunkAction,
  TypedStartListening,
} from "@reduxjs/toolkit";
import { combineSlices, configureStore } from "@reduxjs/toolkit";
import type { MetadataCache, TFile, Vault } from "obsidian";

import { icalRefreshIntervalMillis } from "../constants";
import type DayPlanner from "../main";
import type { DataviewFacade } from "../service/dataview-facade";
import type { PluginData } from "../settings";
import type { ReduxExtraArgument } from "../types";

import { dataviewChange, dataviewSlice } from "./dataview/dataview-slice";
import { globalSlice } from "./global-slice";
import {
  icalRefreshRequested,
  icalSlice,
  type IcalState,
  initialIcalState,
} from "./ical/ical-slice";
import { initListenerMiddleware } from "./listener-middleware";
import { searchSlice } from "./search-slice";
import { settingsSlice } from "./settings-slice";

const rootReducer = combineSlices(
  globalSlice,
  searchSlice,
  dataviewSlice,
  settingsSlice,
  icalSlice,
);

export type RootState = ReturnType<typeof rootReducer>;

export const makeStore = (
  props: Omit<
    ConfigureStoreOptions<RootState>,
    "preloadedState" | "reducer"
  > & {
    preloadedState?: Partial<RootState>;
  },
) => {
  const { preloadedState, middleware } = props;

  return configureStore({
    reducer: rootReducer,
    middleware,
    preloadedState,
  });
};

export function initStoreForPlugin(props: {
  pluginData: PluginData;
  dataviewFacade: DataviewFacade;
  vault: Vault;
  metadataCache: MetadataCache;
  plugin: DayPlanner;
}) {
  const { pluginData, dataviewFacade, vault, metadataCache, plugin } = props;

  const icalStateWithCachedRawIcals: IcalState = {
    ...initialIcalState,
    plainTextIcals: pluginData.rawIcals || [],
  };

  // todo: extra is not needed at all
  const listenerMiddleware = initListenerMiddleware({
    extra: {
      dataviewFacade,
      vault,
      metadataCache,
      onIcalsFetched: async (rawIcals) => {
        await plugin.saveData({ ...plugin.settings(), rawIcals });
      },
    },
  });

  const store = makeStore({
    preloadedState: {
      ical: icalStateWithCachedRawIcals,
    },
    middleware: (getDefaultMiddleware) => {
      return getDefaultMiddleware().concat(listenerMiddleware.middleware);
    },
  });

  plugin.register(() => {
    listenerMiddleware.clearListeners();
  });

  plugin.registerInterval(
    window.setInterval(() => {
      store.dispatch(icalRefreshRequested());
    }, icalRefreshIntervalMillis),
  );

  plugin.registerEvent(
    metadataCache.on(
      // @ts-expect-error
      "dataview:metadata-change",
      (eventType: unknown, file: TFile) =>
        store.dispatch(dataviewChange(file.path)),
    ),
  );

  return { store, listenerMiddleware };
}

export type AppStore = ReturnType<typeof makeStore>;
export type AppDispatch = AppStore["dispatch"];
export type AppThunk<ThunkReturnType = void> = ThunkAction<
  ThunkReturnType,
  RootState,
  unknown,
  Action
>;
export type StartListeningFn = TypedStartListening<
  RootState,
  AppDispatch,
  ReduxExtraArgument
>;
export type AppListenerEffect<A extends Action = Action> = ListenerEffect<
  A,
  RootState,
  AppDispatch
>;
export type AppListenerMiddlewareInstance = ListenerMiddlewareInstance<
  RootState,
  AppDispatch,
  ReduxExtraArgument
>;
