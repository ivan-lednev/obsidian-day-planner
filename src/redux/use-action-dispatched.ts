import type { Action } from "@reduxjs/toolkit";
import { readable } from "svelte/store";

import type { AppListenerMiddlewareInstance } from "./store";

export function useActionDispatched(props: {
  listenerMiddleware: AppListenerMiddlewareInstance;
}) {
  const { listenerMiddleware } = props;

  return readable<Action>(undefined, (set) => {
    return listenerMiddleware.startListening({
      predicate: () => true,
      effect: (action) => {
        set(action);
      },
    });
  });
}
