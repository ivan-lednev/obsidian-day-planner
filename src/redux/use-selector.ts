import type { Selector } from "@reduxjs/toolkit";
import { type Subscriber } from "svelte/store";

import { type AppStore, type RootState } from "./store";

const defaultEqualityFn = <T>(a: T, b: T) => a === b;

// Inspiration: https://gist.github.com/egargan/20267f9d481e9493a9627a8448034b09
export function createUseSelector(store: AppStore) {
  return <T>(
    selector: Selector<RootState, T>,
    props?: {
      equalityFn?: (a: T, b: T) => boolean;
    },
  ) => {
    const { equalityFn = defaultEqualityFn } = props || {};

    return {
      subscribe: (fn: Subscriber<T>) => {
        let lastSelectorValue = selector(store.getState());
        fn(lastSelectorValue);

        return store.subscribe(() => {
          const selectorValue = selector(store.getState());

          if (!equalityFn(selectorValue, lastSelectorValue)) {
            lastSelectorValue = selectorValue;
            fn(lastSelectorValue);
          }
        });
      },
    };
  };
}

export type UseSelector = ReturnType<typeof createUseSelector>;
