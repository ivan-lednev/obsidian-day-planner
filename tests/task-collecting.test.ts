import { readdir, readFile, stat } from "node:fs/promises";
import path from "node:path";

import { isAnyOf } from "@reduxjs/toolkit";
import { MetadataCache, type Vault } from "obsidian";
import { derived, get, writable } from "svelte/store";
import { describe, expect, test, vi } from "vitest";

import {
  dataviewChange,
  listPropsParsed,
  selectDataviewLoaded,
  selectListProps,
} from "../src/redux/dataview/dataview-slice";
import { editCanceled, initialState } from "../src/redux/global-slice";
import { selectRemoteTasks } from "../src/redux/ical/ical-slice";
import { initListenerMiddleware } from "../src/redux/listener-middleware";
import { selectDataviewSource } from "../src/redux/settings-slice";
import { makeStore, type RootState } from "../src/redux/store";
import { useActionDispatched } from "../src/redux/use-action-dispatched";
import { createUseSelector } from "../src/redux/use-selector";
import type { DataviewFacade } from "../src/service/dataview-facade";
import type { WorkspaceFacade } from "../src/service/workspace-facade";
import { defaultSettingsForTests } from "../src/settings";
import type { PointerDateTime } from "../src/types";
import { useTasks } from "../src/ui/hooks/use-tasks";
import { getUpdateTrigger } from "../src/util/store";

import {
  createInMemoryFile,
  FakeDataviewFacade,
  FakeMetadataCache,
  FakePeriodicNotes,
  FakeWorkspaceFacade,
  getPathToDiff,
  InMemoryVault,
} from "./test-utils";

import type { PeriodicNotes } from "../src/service/periodic-notes";
import { isNotVoid } from "typed-assert";
import { EditMode } from "../src/ui/hooks/use-edit/types";
import type { Moment } from "moment";
import { noop } from "lodash/fp";
import { TransactionWriter } from "../src/service/diff-writer";
import { createUpdateHandler } from "../src/create-update-handler";
import { VaultFacade } from "../src/service/vault-facade";
import { isLocal, type Task } from "../src/task-types";
import { getOneLineSummary } from "../src/util/task-utils";

const { join } = path.posix;

const defaultVisibleDays = ["2024-09-26"];

const defaultPreloadedStateForTests: Partial<RootState> = {
  obsidian: {
    ...initialState,
    visibleDays: defaultVisibleDays,
  },
  settings: {
    settings: {
      ...defaultSettingsForTests,
    },
  },
};

function makeStoreForTests(props: {
  vault: Vault;
  metadataCache: MetadataCache;
  dataviewFacade: DataviewFacade;
  preloadedState?: Partial<RootState>;
}) {
  const {
    metadataCache,
    dataviewFacade,
    vault,
    preloadedState = defaultPreloadedStateForTests,
  } = props;

  const listenerMiddleware = initListenerMiddleware({
    extra: {
      metadataCache,
      dataviewFacade,
      vault,
      onIcalsFetched: async () => {},
    },
  });

  const store = makeStore({
    preloadedState,
    middleware: (getDefaultMiddleware) => {
      return getDefaultMiddleware().concat(listenerMiddleware.middleware);
    },
  });

  return { store, listenerMiddleware };
}

/**
 * This assembles all the system separated from the UI, Obsidian, time, DOM, etc.
 * @param props
 */
async function setUp(props: { visibleDays?: string[] }) {
  const { visibleDays = defaultVisibleDays } = props;

  const fixturesDirPath = "fixtures";
  const dumpPath = join(fixturesDirPath, "metadata-dump", "tasks.json");
  const fixtureVaultPath = join(fixturesDirPath, "fixture-vault");

  const files = await readdir(fixtureVaultPath);

  const inMemoryFiles = await Promise.all(
    files.map(async (file) => {
      const filePath = join(fixtureVaultPath, file);

      const stats = await stat(filePath);

      if (!stats.isFile()) {
        throw new TypeError(
          `Only files are supported in fixture vault, not this: ${filePath}`,
        );
      }

      const contents = await readFile(filePath, "utf8");

      return createInMemoryFile({ path: filePath, contents });
    }),
  );

  const rawMetadataDump = await readFile(dumpPath, "utf-8");

  const metadataDump = JSON.parse(rawMetadataDump);
  const { tasks, lists, cachedMetadata } = metadataDump;

  const pathToInMemoryFile = Object.fromEntries(
    inMemoryFiles.map((it) => [it.path, it]),
  );

  const dailyNoteFileNames = ["2025-07-19", "2025-07-20"];

  const inMemoryDailyNotes = dailyNoteFileNames.map((it) => {
    const path = join(fixtureVaultPath, `${it}.md`);
    const file = pathToInMemoryFile[path];

    isNotVoid(file, `There is no file for key: '${it}'`);

    return { path, file, date: window.moment(it) };
  });

  const periodicNotes = new FakePeriodicNotes(
    inMemoryDailyNotes,
  ) as unknown as PeriodicNotes;

  const dataviewFacade = new FakeDataviewFacade({
    tasks,
    lists,
  }) as unknown as DataviewFacade;

  const metadataCache = new FakeMetadataCache(
    cachedMetadata,
  ) as unknown as MetadataCache;

  const vault = new InMemoryVault(inMemoryFiles);

  const getTasksApi = () => {
    throw new Error("Can't access tasks API inside tests");
  };

  const vaultFacade = new VaultFacade(vault as unknown as Vault, getTasksApi);

  const transactionWriter = new TransactionWriter(vaultFacade);

  const workspaceFacade =
    new FakeWorkspaceFacade() as unknown as WorkspaceFacade;

  const {
    store,
    store: { dispatch, getState },
    listenerMiddleware,
  } = makeStoreForTests({
    dataviewFacade,
    metadataCache,
    vault: vault as unknown as Vault,
    preloadedState: {
      ...defaultPreloadedStateForTests,
      obsidian: {
        ...initialState,
        visibleDays,
      },
    },
  });

  inMemoryFiles.forEach(({ path }) => {
    store.dispatch(dataviewChange(path));
  });

  const visibleDaysStore = writable(visibleDays.map((it) => window.moment(it)));
  const isOnline = writable(true);
  const layoutReady = writable(true);
  const settingsStore = writable(defaultSettingsForTests);
  const currentTime = writable(window.moment());

  const onEditCanceled = vi.fn();
  const onEditConfirmed = vi.fn();

  const onUpdate = createUpdateHandler({
    settings: () => defaultSettingsForTests,
    transactionWriter,
    vaultFacade,
    periodicNotes,
    onEditCanceled,
    onEditConfirmed,
    getTextInput: () => Promise.resolve("Text input"),
    getConfirmationInput: () => Promise.resolve(true),
  });

  // todo: remove copy-pasta
  const useSelector = createUseSelector(store);
  const actionDispatched = useActionDispatched({ listenerMiddleware });

  const remoteTasks = useSelector(selectRemoteTasks);
  const listProps = useSelector(selectListProps);
  const dataviewLoaded = useSelector(selectDataviewLoaded);
  const dataviewSource = useSelector(selectDataviewSource);

  const dataviewRefreshSignal = derived(
    actionDispatched,
    ($actionDispatched, set) => {
      // todo: fix in real code
      if (isAnyOf(listPropsParsed, editCanceled)($actionDispatched)) {
        set($actionDispatched);
      }
    },
  );

  const taskUpdateTrigger = derived(
    [dataviewRefreshSignal, dataviewSource],
    getUpdateTrigger,
  );

  const pointerDateTime = writable<PointerDateTime>({
    dateTime: window.moment(),
    type: "dateTime",
  });
  // ---

  const {
    tasksWithActiveClockProps,
    getDisplayedTasksWithClocksForTimeline,
    tasksWithTimeForToday,
    editContext,
    newlyStartedTasks,
  } = useTasks({
    onUpdate,
    onEditAborted: () => {},
    periodicNotes,
    dataviewFacade,
    metadataCache,
    workspaceFacade,
    isOnline,
    visibleDays: visibleDaysStore,
    layoutReady,
    debouncedTaskUpdateTrigger: taskUpdateTrigger,
    dataviewChange: taskUpdateTrigger,
    settingsStore,
    currentTime,
    pointerDateTime,
    remoteTasks,
    listProps,
  });

  const allTasks = derived(
    editContext.dayToDisplayedTasks,
    ($dayToDisplayedTasks) => {
      return Object.values($dayToDisplayedTasks).flatMap(
        ({ withTime, noTime }) => withTime.concat(noTime),
      );
    },
  );

  // this prevents the store from resetting;
  allTasks.subscribe(noop);

  // todo: copy-pasta
  function moveCursorTo(
    dateTime: Moment,
    type: "date" | "dateTime" = "dateTime",
  ) {
    pointerDateTime.set({
      dateTime,
      type,
    });
  }

  function findTask(predicate: (task: Task) => boolean) {
    const found = get(allTasks).filter(isLocal).find(predicate);

    isNotVoid(found, "Task not found. Wrong test code");

    return found;
  }

  function findByText(text: string) {
    return findTask((it) => getOneLineSummary(it).includes(text));
  }

  return {
    dispatch,
    getState,
    tasksWithActiveClockProps,
    getDisplayedTasksWithClocksForTimeline,
    tasksWithTimeForToday,
    editContext,
    newlyStartedTasks,
    periodicNotes,
    moveCursorTo,
    onUpdate,
    vault,
    findTask,
    findByText,
    allTasks,
  };
}

describe("Task views", () => {
  describe("Log entry views", () => {
    test("Reads log data", async () => {
      const { getDisplayedTasksWithClocksForTimeline } = await setUp({
        visibleDays: ["2025-07-19"],
      });

      const displayedTasks = getDisplayedTasksWithClocksForTimeline(
        window.moment("2025-07-19"),
      );

      await vi.waitFor(() => expect(get(displayedTasks)).toHaveLength(3));

      expect(get(displayedTasks)).toMatchObject([
        {
          startTime: window.moment("2025-07-19 12:00"),
          durationMinutes: 150,
        },
        {
          startTime: window.moment("2025-07-19 15:00"),
          durationMinutes: 90,
        },
        {
          startTime: window.moment("2025-07-19 13:00"),
          durationMinutes: 210,
        },
      ]);
    });

    test.todo("Ignores invalid dates, extra keys");

    test.todo("Tasks with active clocks receive durations set to current time");

    test.todo("Splits log entries over midnight");
  });

  test("Ignores tasks and lists outside of planner section", async () => {
    const { editContext } = await setUp({
      visibleDays: ["2025-07-19"],
    });

    const displayedTasks = editContext.getDisplayedTasksForTimeline(
      window.moment("2025-07-19"),
    );

    await vi.waitFor(() =>
      expect(get(displayedTasks)?.withTime.length).toBeGreaterThan(0),
    );

    expect(get(displayedTasks)?.withTime).toHaveLength(1);
    expect(get(displayedTasks)?.withTime).toMatchObject([
      { startTime: window.moment("2025-07-19 11:00"), durationMinutes: 30 },
    ]);
  });

  test.todo("Tasks do not contain duplicates");

  test.todo("Combines tasks from daily notes with tasks from other files");

  describe("Editing", () => {
    test("Edits tasks", async () => {
      const { editContext, moveCursorTo, vault, findByText } = await setUp({
        visibleDays: ["2025-07-19"],
      });

      const displayedTasks = editContext.getDisplayedTasksForTimeline(
        window.moment("2025-07-19"),
      );

      await vi.waitFor(() =>
        expect(get(displayedTasks)?.withTime.length).toBeGreaterThan(0),
      );

      editContext.handlers.handleGripMouseDown(
        findByText("List item under planner heading"),
        EditMode.DRAG,
      );

      moveCursorTo(window.moment("2025-07-19 17:00"));

      await editContext.confirmEdit();

      expect(getPathToDiff(vault.initialState, vault.state)).toMatchSnapshot();
    });

    test("Moves tasks between daily notes", async () => {
      const { editContext, moveCursorTo, vault, findByText, allTasks } =
        await setUp({
          visibleDays: ["2025-07-19"],
        });

      await vi.waitFor(() => expect(get(allTasks).length).toBeGreaterThan(0));

      editContext.handlers.handleGripMouseDown(
        findByText("List item under planner heading"),
        EditMode.DRAG,
      );

      moveCursorTo(window.moment("2025-07-20 17:00"));

      await editContext.confirmEdit();

      expect(getPathToDiff(vault.initialState, vault.state)).toMatchSnapshot();
    });
  });

  describe("Frontmatter", () => {
    test.todo("Shows log entries from frontmatter");
  });
});
