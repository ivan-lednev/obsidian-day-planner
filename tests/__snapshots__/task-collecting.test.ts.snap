// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task views > Editing > Edits tasks 1`] = `
{
  "fixtures/fixture-vault/2025-07-19.md": "
- - 11:00 - 11:30 List item under planner heading
+ - 17:00 - 17:30 List item under planner heading
",
}
`;

exports[`Task views > Editing > Moves tasks between daily notes 1`] = `
{
  "fixtures/fixture-vault/2025-07-19.md": "
- - 11:00 - 11:30 List item under planner heading
",
  "fixtures/fixture-vault/2025-07-20.md": "
+ # Day planner
+ 
+ - 17:00 - 17:30 List item under planner heading
",
}
`;
